postprocess/postprocess.go

```diff
@@ -0,0 +1,230 @@
+
+	// ErrXNotFloat64 denotes that an X value is not a float64.
+	ErrXNotFloat64 = errors.New("could not parse provided X as a float64")
+
+	// ErrYNotProvided denotes that a Y value was not provided.
+	ErrYNotProvided = errors.New("could X not provided")
+
+	// ErrYNotFloat64 denotes that an Y value is not a float64.
+	ErrYNotFloat64 = errors.New("could not parse provided X as a float64")
+
+	// ErrRemovingPoints denotes that something unexpected happened during removal.
+	ErrRemovingPoints = errors.New("unexpected number of points after removal")
+)
+
+// Task can be used to construct a postprocessing step.
+type Task struct {
+	Instruction Instruction
+	Points      []r3.Vector
+}
+
+// ParseDoCommand parses postprocessing DoCommands into Tasks.
+func ParseDoCommand(
+	unstructuredPoints interface{},
+	instruction Instruction,
+) (Task, error) {
+	pointSlice, ok := unstructuredPoints.([]interface{})
+	if !ok {
+		return Task{}, ErrPointsNotASlice
+	}
+
+	task := Task{}
+	for _, point := range pointSlice {
+		pointMap, ok := point.(map[string]interface{})
+		if !ok {
+			return Task{}, ErrPointNotAMap
+		}
+
+		x, ok := pointMap[xKey]
+		if !ok {
+			return Task{}, ErrXNotProvided
+		}
+
+		xFloat, ok := x.(float64)
+		if !ok {
+			return Task{}, ErrXNotFloat64
+		}
+
+		y, ok := pointMap[yKey]
+		if !ok {
+			return Task{}, ErrYNotProvided
+		}
+
+		yFloat, ok := y.(float64)
+		if !ok {
+			return Task{}, ErrXNotFloat64
+		}
+
+		task.Points = append(task.Points, r3.Vector{X: xFloat, Y: yFloat})
+	}
+	return task, nil
+}
+
+/*
+UpdatePointCloud iterated through a list of tasks and adds or removes points from data
+and writes the updated pointcloud to updatedData.
+*/
+func UpdatePointCloud(
+	data []byte,
+	updatedData *[]byte,
+	tasks []Task,
+) error {
+	*updatedData = append(*updatedData, data...)
+
+	// iterate through tasks and add or remove points
+	for _, task := range tasks {
+		switch task.Instruction {
+		case Add:
+			err := updatePointCloudWithAddedPoints(updatedData, task.Points)
+			if err != nil {
+				return err
+			}
+		case Remove:
+			err := updatePointCloudWithRemovedPoints(updatedData, task.Points)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func updatePointCloudWithAddedPoints(updatedData *[]byte, points []r3.Vector) error {
+	reader := bytes.NewReader(*updatedData)
+	pc, err := pointcloud.ReadPCD(reader)
+	if err != nil {
+		return err
+	}
+
+	for _, point := range points {
+		/*
+			Viam expects pointcloud data with fields "x y z" or "x y z rgb", and for
+			this to be specified in the pointcloud header in the FIELDS entry. If color
+			data is included in the pointcloud, Viam's services assume that the color
+			value encodes a confidence score for that data point. Viam expects the
+			confidence score to be encoded in the blue parameter of the RGB value, on a
+			scale from 1-100.
+		*/
+		err := pc.Set(point, pointcloud.NewColoredData(color.NRGBA{B: fullConfidence}))
```

Comment:
[q]Would we be interested in adding another color here to mark the pointcloud as an added obstacle? 

the prime component currently overwrites all colors to be grayscale based on the blue channel, so actually visualizing the obstacles would require a change in the prime component to handle the new channel but I think it would be a cool visual

Reply:
Will it be more than a day's work? If so yep. If it's complicated, we'll do in a future date when we more formally do this. Save this idea, it's good.